pysdk/trading/service_bot.py

```diff
@@ -0,0 +1,200 @@
+        process.wait()
+        data, err = process.communicate()
+        if process.returncode is 0:
+            nibid_exchange_rates = data.decode('utf-8')
+            nibid_exchange_rates_json = json.loads(nibid_exchange_rates)  
+            if len(nibid_exchange_rates_json["exchange_rates"]) != 0: 
+                index_quote = float(nibid_exchange_rates_json["exchange_rates"][0]["exchange_rate"])
+            else: 
+                self.pos_dict["errors"].append("Exhange Rates unavailable")
+                index_quote = 0
+            index_quote = 25000
+            print("Index Quote: ",index_quote)
+        else:
+            self.pos_dict["errors"].append(err)
+        return index_quote
+    
+    # Fetch the mark, price_mult, quote reserves, and market bias from chain
+    def find_mark_quote(self) -> Tuple[float, float, float, float]:
+        process = subprocess.Popen("nibid q perp markets", shell=True, stdout=subprocess.PIPE)
+        process.wait()
+        data, err = process.communicate()
+        if process.returncode is 0:
+            nibid_markets = data.decode('utf-8')
+            nibid_markets_json = json.loads(nibid_markets)
+            nibid_amm_json = nibid_markets_json["amm_markets"][0]
+            quote_asset_reserve = float(nibid_amm_json['amm']['quote_reserve'])
+            base_asset_reserve = float(nibid_amm_json['amm']['base_reserve'])
+            price_mult = float(nibid_amm_json['amm']['price_multiplier'])
+
+            net_position = float(nibid_amm_json['amm']['total_long']) - float(nibid_amm_json['amm']['total_short'])
+
+            mark_quote = base_asset_reserve * price_mult/quote_asset_reserve
+            print("Mark Quote: ",mark_quote)
+        else:
+            self.pos_dict["errors"].append(err)
+        return mark_quote, quote_asset_reserve, net_position, price_mult
+    
+    # Check if the position is against the market based on mark & index
+    def is_pos_against_market(self, index, mark) -> bool:
+        if mark>index:
+            if(self.isLong) == True:
+                against_market = True
+            else:
+                against_market = False
+        if index > mark:
+            if(self.isLong) == True:
+                against_market = False
+            else:
+                against_market = True
+        return against_market
+        
+        
+    # Calculate the quote needed to move the price to reach the target price
+    def quote_needed_to_move_price(self, current_price:float, target_price:float, quote_reserve:float) -> float: 
+        qp = target_price / current_price
+        return -(quote_reserve / math.sqrt(qp) - quote_reserve)  
+
+
+def main():
+    # Instantiate the TradingBot class within the main() function
+    while True:
+        bot = TradingBot()
+
+        mark_quote, quote_asset_reserve, net_position, price_multiplier = bot.find_mark_quote()
+        index_quote = bot.find_index_quote()
```

Comment:
I'd extract the logic from this `while` loop into a method on the `TradingBot` called `run`. Then, references to `bot` will just be `self`